src/commonMain/kotlin/fr/acinq/lightning/io/Peer.kt

```diff
@@ -200,26 +202,38 @@ class Peer(
         }
         launch {
             finalWallet.walletStateFlow
-                .distinctUntilChangedBy { Pair(it.unconfirmedBalance, it.confirmedBalance) }
+                .distinctUntilChangedBy { it.totalBalance }
                 .collect { wallet ->
-                    logger.info { "${wallet.totalBalance} available on final wallet (${wallet.unconfirmedBalance} unconfirmed)" }
+                    logger.info { "${wallet.totalBalance} available on final wallet with ${wallet.utxos.size} utxos" }
                 }
         }
         launch {
             swapInWallet.walletStateFlow
                 .filter { it.consistent }
                 .fold(emptySet<WalletState.Utxo>()) { reservedUtxos, wallet ->
-                    // reservedUtxos are part of a previously issued RequestChannelOpen command
-                    val availableWallet = wallet.minus(reservedUtxos)
-                    val balance = availableWallet.confirmedBalance
-                    logger.info { "swap-in wallet balance: confirmed=$balance, unconfirmed=${availableWallet.unconfirmedBalance}" }
-                    if (balance > 0.sat) {
-                        logger.info { "swap-in wallet: requesting channel using confirmed balance: $balance" }
-                        input.send(RequestChannelOpen(Lightning.randomBytes32(), availableWallet))
-                        reservedUtxos.union(availableWallet.confirmedUtxos)
-                    } else {
-                        reservedUtxos
-                    }.intersect(wallet.utxos.toSet()) // drop utxos no longer in wallet
+                    when (val currentBlockHeight = currentTipFlow.value?.first) {
```

Comment:
I'd rather remove this `when` and simply do `val currentBlockHeight = currentTipFlow.filterNotNull().first().first`. It's simpler, consistent with the rest of the file, and that case will never happen anyway since if we receive a balance event we have to be connected in the first place.

Reply:
We can't, the compiler won't let me do that because `.first()` needs a `suspend`...but if you find the right way to do this, that's what I initially wanted to write!

Reply:
This code is already inside a coroutine so the compiler shouldn't give you a hard time.

The following compiles on my machine™:

```kotlin
        launch {
            swapInWallet.walletStateFlow.filter { it.consistent }.fold(emptySet<WalletState.Utxo>()) { reservedUtxos, wallet ->
                    val currentBlockHeight = currentTipFlow.filterNotNull().first().first
                    // reservedUtxos are part of a previously issued RequestChannelOpen command
                    val availableWallet = wallet.minus(reservedUtxos).withConfirmations(currentBlockHeight, walletParams.swapInConfirmations)
                    logger.info { "swap-in wallet balance: deeplyConfirmed=${availableWallet.deeplyConfirmed.balance}, weaklyConfirmed=${availableWallet.weaklyConfirmed.balance}, unconfirmed=${availableWallet.unconfirmed.balance}" }
                    val utxos = when {
                        // When migrating from the legacy android app, we use all utxos, even unconfirmed ones.
                        isMigrationFromLegacyApp -> availableWallet.all
                        else -> availableWallet.deeplyConfirmed
                    }
                    if (utxos.balance > 0.sat) {
                        logger.info { "swap-in wallet: requesting channel using ${utxos.size} utxos with balance=${utxos.balance}" }
                        input.send(RequestChannelOpen(Lightning.randomBytes32(), utxos))
                        reservedUtxos.union(utxos.utxos)
                    } else {
                        reservedUtxos
                    }.intersect(wallet.utxos.toSet()) // drop utxos no longer in wallet
                }
        }
```

Reply:
I don't know what I previously did, that's exactly what I tried but the compiler was complaining...I must have gotten it wrong, fixed in https://github.com/ACINQ/lightning-kmp/pull/468/commits/e9dbdbcc3d4b5c602d27de20e24cf066c64b70aa